### PR TITLE
Emit an event to indicate a successful acknowledgement in the ICA module

### PR DESCRIPTION
**Pull request by chatton**
## Description

This PR updates the ICA module to emit an event on success in `OnRecvPacket`

closes: #1452 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: modules/apps/27-interchain-accounts/host/keeper/events.go [hunk 1]
```diff
@@ -18,3 +18,14 @@ func EmitWriteErrorAcknowledgementEvent(ctx sdk.Context, packet exported.PacketI
...
 	)
 }
+
+// EmitSuccessfulAcknowledgementEvent emits an event signalling a successful acknowledgement
+func EmitSuccessfulAcknowledgementEvent(ctx sdk.Context, packet exported.PacketI) {
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			icatypes.EventTypePacket,
+			sdk.NewAttribute(sdk.AttributeKeyModule, icatypes.ModuleName),
+			sdk.NewAttribute(icatypes.AttributeKeyHostChannelID, packet.GetDestChannel()),
```
Comment (chatton) [inline modules/apps/27-interchain-accounts/host/keeper/events.go:28]: I would like input on whether or not there are additional fields/attributes that should be included in this event. Please let me know if I've missed anything!

Comment (crodriguezvega) [reply modules/apps/27-interchain-accounts/host/keeper/events.go:28]: This looks fine, maybe we could add a `AttributeKeyAckSuccess` [like for transfer](https://github.com/cosmos/ibc-go/blob/25767f6bdb5bab2c2a116b41d92d753c93e18121/modules/apps/transfer/ibc_module.go#L202).

But I am also now wondering if we could merge both event functions into one and have something like this:

```
ctx.EventManager().EmitEvent(
	sdk.NewEvent(
		icatypes.EventTypePacket,
		sdk.NewAttribute(sdk.AttributeKeyModule, icatypes.ModuleName),
		sdk.NewAttribute(icatypes.AttributeKeyAckError, err.Error()),
		sdk.NewAttribute(icatypes.AttributeKeyHostChannelID, packet.GetDestChannel()),
                sdk.NewAttribute(icatypes.AttributeKeyAckSuccess, fmt.Sprintf("%t", ack.Success())),
	),
)
```

with the `icatypes.AttributeKeyAckError` empty if it was successful.

Comment (chatton) [reply modules/apps/27-interchain-accounts/host/keeper/events.go:28]: I was actually thinking something very similar with regards to merging the functions.

File: modules/apps/27-interchain-accounts/host/ibc_module.go [hunk 2]
```diff
@@ -110,15 +110,21 @@ func (im IBCModule) OnRecvPacket(
...
+		ack = types.NewErrorAcknowledgement(err)
+	}
 
-		return types.NewErrorAcknowledgement(err)
+	if err == nil {
+		// Emit an event indicating a successful acknowledgement.
+		keeper.EmitSuccessfulAcknowledgementEvent(ctx, packet)
+	} else {
+		// Emit an event including the error msg.
+		keeper.EmitWriteErrorAcknowledgementEvent(ctx, packet, err)
```
[the remark this answers is not in the document]
Comment (chatton) [reply modules/apps/27-interchain-accounts/host/ibc_module.go:123]: My intention was to keep the style similar to what we have [here](https://github.com/cosmos/ibc-go/blob/f05a7cf8840ead622e7b5628466e1f9a311a717d/modules/apps/transfer/ibc_module.go#L190)

File: CHANGELOG.md [hunk 3]
```diff
@@ -125,6 +125,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * (channel) [\#644](https://github.com/cosmos/ibc-go/pull/644) Adds `GetChannelConnection` to the ChannelKeeper. This function returns the connectionID and connection state associated with a channel. 
 * (channel) [\#647](https://github.com/cosmos/ibc-go/pull/647) Reorganizes channel handshake handling to set channel state after IBC application callbacks. 
 * (client) [\#724](https://github.com/cosmos/ibc-go/pull/724) `IsRevisionFormat` and `IsClientIDFormat` have been updated to disallow newlines before the dash used to separate the chainID and revision number, and the client type and client sequence. 
+* (interchain-accounts) [\#1466](https://github.com/cosmos/ibc-go/pull/1466) Emit event when there is an acknowledgement error during `OnRecvPacket`.
```
Comment (damiannolan) [inline CHANGELOG.md:128]: Should this indicate an acknowledgement event is emitted for both success and error cases?

File: modules/apps/27-interchain-accounts/host/keeper/events.go [hunk 4]
```diff
@@ -1,20 +1,29 @@
...
 	icatypes "github.com/cosmos/ibc-go/v3/modules/apps/27-interchain-accounts/types"
 	"github.com/cosmos/ibc-go/v3/modules/core/exported"
 )
 
-// EmitWriteErrorAcknowledgementEvent emits an event signalling an error acknowledgement and including the error details
-func EmitWriteErrorAcknowledgementEvent(ctx sdk.Context, packet exported.PacketI, err error) {
+// EmitAcknowledgementEvent emits an event signalling a successful or failed acknowledgement and including the error
+// details if any.
+func EmitAcknowledgementEvent(ctx sdk.Context, packet exported.PacketI, ack exported.Acknowledgement, err error) {
+	errorMsg := ""
```
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/host/keeper/events.go:15]: nit: should we favour a var declaration here? 

```suggestion
	var errorMsg string
```

File: modules/apps/27-interchain-accounts/host/keeper/events.go [hunk 5]
```diff
@@ -1,20 +1,29 @@
 package keeper
 
 import (
+	"fmt"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/v3/modules/apps/transfer/types"
```
Comment (damiannolan) [inline modules/apps/27-interchain-accounts/host/keeper/events.go:6]: I see this is imported from `transfer` for `AttributeKeyAckSuccess`. Could we redefine this in the `interchain-accounts` module package `types` so we don't have apps depending directly on one another unnecessarily?

**Review by damiannolan — COMMENTED** [review]
Just left a few very minor comments

**Pull request merged**